### ISSUE TITLE
How to make the decision

### ISSUE THREAD

**Issue by jimmyhmiller** [trabian/clojure-elixir-comparison#3]
We need to have some process down about how this decision will be made. This is me writing down the methodology I think we should follow in making this decision, if there is any points that you feel need changed, let me know.

There are two types of considerations that come into choosing a language, subjective and objective. Often times people consider the subjective to be less imporantant than the objective. That is something I think we should avoid. Subjective considerations (taste) play a large part in software development. That being said, subjective and objective considerations do differ.

Subjective tastes are person relative. When we are considering the subjective, we should look for broad agreement, rather than individual preference. For example, our team as a whole has a strong preference towards functional programming, hence Java is not one of the languages we are considering. Subjective considerations work best when there is consensus.

Objective considerations apply across people. Given adequate evidence, objective facts should be agreed on by all members of the team. Many objective statements are common knowledge. For instance, the fact that Elixir runs on the Erlang VM does not require a citation unless this is knowledge a team member does not share. Some objective facts however, require careful definition and sufficient evidence. For instance, X is more scalable than Y requires some way in which we can measure scalability and some sufficent test to prove such a statement. 

The objective and subjective connect as well. It is our subjective tastes that let us weigh which objective facts we consider important. For example, Java supports object orientation is something no one on our team would dispute, but this is subjectively not seen as a positive feature. Even positive objective facts about a language will be weighed differently.

**TL;DR** - Give objective and subjective reasons.

## The final choice

Regardless of how this process plays out, I think it makes sense to have a method for how we will choose at the end. My personal opinion is that the language we use for this project should be chosen by vote. This way everyone has an equal voice in the decision.

**Comment by trabianmatt** [comment]
Thanks for laying this out. I'm fine with a vote for this project, though as always we should keep our options for the future.

The primary subjective concern I have with Clojure right now is its lack of proscription. I've built Clojure(script) applications in the ~1k LOC range, so nothing big and nothing that's in production, but also not trivial. The challenge that I ran into is that there were so many ways of doing something that I had to filter through the range of possibilities every time I needed to make a decision. Perhaps this is the "rule of least power" or perhaps it was a sign of personal inexperience, but I've also tried to learn from the experience of others and it seems to be shared by at least some subset of the population.

I can find my own patterns if I'm the only one working on it. The concern is whether it will add friction around "what are Trabian's patterns?". Is there a sufficiently idiomatic approach to Clojure now that will serve as a guide? Will new developers be able to work productively within our environment? Will our future selves be able to do the same?

Again, I like Clojure. I like Clojurescript. My primary evaluation question (because I know that we can build solid, performant systems in either platform) is in our ability to either grow the team productively or be so productive with the smaller team that it's less of an issue. And in general, in which language will a system be most straightforward to come back in and understand what's going on?

**Comment by jimmyhmiller** [comment]
>The primary subjective concern I have with Clojure right now is its lack of proscription...The challenge that I ran into is that there were so many ways of doing something that I had to filter through the range of possibilities every time I needed to make a decision. 

I find this interesting that you are worried about this. This seems to me to perfectly describe the Javascript ecosystem. It even describes the React ecosystem.

>The concern is whether it will add friction around "what are Trabian's patterns?"

There will of course be a learning curve with these things. As there was with React, as there would be in any language we could use. Each of our React apps have looked incredibly different in structure. 

That being said, I don't think we will find any difficulty unique to Clojure in this area, in fact I think it has a number of tools that help with 1) establishing practices and 2) enabling productivity for people unfamiliar with the code base.

One tool I think is invaluable for discovery and understanding in a new code base is Clojure's ability to do live programming. Being able to repl into a running program, modify it and inspect values is incredibly valuable. Because of this feature, I've been able to make changes to codebases I am completely unfamiliar with.

Clojure's functional approach and immutable data structures eliminate much of the complexity of understanding existing systems. Of course, these are features that Elixir shares, but it doesn't (in my opinion) take full advantage of them. Logic is split up with message passing, mimicking the problem that oop has, each object (process) has its own little api (language) you have to learn. This emphasis on functional programming and the importance of these data structures makes a huge difference in understand code.

Finally, Clojure.spec while not fully released offers a very promising approach to embedding those best practices.The instrumentation that specs provide help developers not make mistakes. They can automatically produce tests to ensure that things are working the way we expect.

### Conclusion

All that being said, I think Clojure's lack of proscription is its best feature. Clojure's ecosystem is like that of react's, small building blocks to make your final solution. This is a fantastic way to develop rather than trying to fit your program into some predefined framework. This bottom up approach to development makes your programs more habitable, more flexible.

**Comment by jimmyhmiller** [comment]
Nothing actionable here. Closing
